src/app/ui/toolbar.cpp

```diff
@@ -616,6 +675,95 @@ void ToolBar::drawToolIcon(Graphics* g, int group_index, SkinPartPtr skin, os::S
   }
 }
 
+int ToolBar::getHiddenGroups()
+{
+  const int height = size().h;
+  if (height < m_minHeight) {
+    int hidden = (m_minHeight - height) / getToolIconSize(this).h;
+    if (hidden >= 1)
+      return std::clamp(hidden + 1, 2, App::instance()->toolBox()->getGroupsCount() - 1);
+  }
+
+  return 0;
+}
+
+void ToolBar::showHiddenToolPopup()
+{
+  if (m_popupWindow) {
+    if (m_closeConn)
+      m_closeConn.disconnect();
+
+    onClosePopup();
+    closePopupWindow();
+  }
+
+  closeTipWindow();
+
+  auto toolbox = App::instance()->toolBox();
+  m_popupWindow = new TransparentPopupWindow(PopupWindow::ClickBehavior::CloseOnClickOutsideHotRegion);
+  m_closeConn = m_popupWindow->Close.connect([this] { onClosePopup(); });
+
+  std::vector<ButtonSet::Item*> items;
+  auto toolIconSize = getToolIconSize(this);
+  auto theme = SkinTheme::get(this);
+  ToolGroupList::iterator it = toolbox->begin_group();
+
+  for (int i = 0; i < toolbox->getGroupsCount(); ++i, ++it) {
+    if (i < toolbox->getGroupsCount() - getHiddenGroups())
+      continue;
+
+    ToolGroup* tool_group = *it;
+    for (ToolIterator bit = toolbox->begin(); bit != toolbox->end(); ++bit) {
+      Tool* tool = *bit;
+      if (tool->getGroup() == tool_group) {
+        auto item = new ButtonSet::Item();
+        // TODO: Tooltip with tool->getText(), implement in ButtonSet?
+        item->setIcon(theme->getToolPart(tool->getId().c_str()));
+        item->setId(tool->getId().c_str());
+        item->setMaxSize(toolIconSize);  // TODO: Need a better way to enforce the button/icon size, this one doesn't look great
```

Comment:
@dacap Is there a way to constrain the icon size here without changing how ButtonSet renders things?
![image](https://github.com/user-attachments/assets/3dd51616-7776-4d4d-8c5a-88c612f51071)

It doesn't look super great with some of the icons, but the default size without specifying anything is massive:
![image](https://github.com/user-attachments/assets/836eeccb-bc9f-4d63-bf41-e83de1217d21)